pilot/pkg/networking/core/v1alpha3/envoyfilter/cluster_patch.go

```diff
@@ -25,39 +27,182 @@ import (
 	"istio.io/pkg/log"
 )
 
-func ApplyClusterMerge(pctx networking.EnvoyFilter_PatchContext, efw *model.EnvoyFilterWrapper, c *cluster.Cluster) (out *cluster.Cluster) {
-	defer runtime.HandleCrash(runtime.LogPanic, func(interface{}) {
-		log.Errorf("clusters patch caused panic, so the patches did not take effect")
-	})
-	// In case the patches cause panic, use the clusters generated before to reduce the influence.
-	out = c
+const (
+	RemovePatch = "-"
+	MergeAny    = "*"
+	MergeOne    = "+"
+	Service     = "service"
+	Subset      = "subset"
+	Port        = "port"
+)
+
+func GenerateMatchMap(pctx networking.EnvoyFilter_PatchContext, efw *model.EnvoyFilterWrapper) (cpw map[string][]*model.EnvoyFilterConfigPatchWrapper,
+	serviceMap map[string][]*model.EnvoyFilterConfigPatchWrapper,
+	subsetMap map[string][]*model.EnvoyFilterConfigPatchWrapper,
+	portMap map[string][]*model.EnvoyFilterConfigPatchWrapper) {
+
+	cpw = make(map[string][]*model.EnvoyFilterConfigPatchWrapper)
+	serviceMap = make(map[string][]*model.EnvoyFilterConfigPatchWrapper)
+	subsetMap = make(map[string][]*model.EnvoyFilterConfigPatchWrapper)
+	portMap = make(map[string][]*model.EnvoyFilterConfigPatchWrapper)
 	if efw == nil {
-		return
+		return cpw, serviceMap, subsetMap, portMap
 	}
 	for _, cp := range efw.Patches[networking.EnvoyFilter_CLUSTER] {
-		if cp.Operation != networking.EnvoyFilter_Patch_MERGE {
+		if cp.Operation != networking.EnvoyFilter_Patch_REMOVE &&
+			cp.Operation != networking.EnvoyFilter_Patch_MERGE {
+			continue
+		}
+		if !commonConditionMatch(pctx, cp) {
 			continue
 		}
-		if commonConditionMatch(pctx, cp) && clusterMatch(c, cp) {
+
+		if cp.Match.GetCluster() == nil {
+			if cp.Operation == networking.EnvoyFilter_Patch_REMOVE {
+				cpw[RemovePatch] = append(cpw[RemovePatch], cp)
+				continue
+			}
+			cpw[MergeAny] = append(cpw[MergeAny], cp)
+			continue
+		}
```

Comment:
why do need to handle non cluster patches?

Reply:
when EnvoyFilter_ClusterMatch is nil, it means that all the clusters should be patched. I build the map of filters that should merge or remove any cluster.